content/ecs_advance_schedule/code_review.md

```diff
@@ -0,0 +1,294 @@
+```
+
+#### 3.4 ARM based task deployment code
+
+For the ARM based capacity provider in the ECS cluster, there are quite a few resources that have to be created to start Container resources to run ARM based tasks. Those resources are the ECS Cluster, AutoScaling Group, Capacity Provider and the Task Definition with ARM based docker image configuration and PlacementConstraints configuration.  
+
+```
+Resources:
+# Shared ECS Cluster. 
+  ECSCluster:
+    Type: AWS::ECS::Cluster
+
+# ARM64 based Launch Configuration. 
+  ArmASGLaunchConfiguration:
+    Type: AWS::AutoScaling::LaunchConfiguration
+    Properties:
+      ImageId: !Ref ArmLatestAmiId
+      SecurityGroups: !Split
+        - ','
+        - Fn::ImportValue: !Sub "${VPCStackParameter}-SecurityGroups"
+      InstanceType: !Ref 'ArmInstanceType'
+      IamInstanceProfile: !Ref 'EC2InstanceProfile'
+      UserData:
+        Fn::Base64: !Sub |
+          #!/bin/bash -xe
+          echo ECS_CLUSTER=${ECSCluster} >> /etc/ecs/ecs.config
+          yum install -y aws-cfn-bootstrap
+          /opt/aws/bin/cfn-signal -e $? --stack ${AWS::StackName} --resource ArmECSAutoScalingGroup --region ${AWS::Region}
+
+ # AutoScalingGroup to launch Container Instances using ARM64 Launch Configuration.  
+  ArmECSAutoScalingGroup:
+    Type: AWS::AutoScaling::AutoScalingGroup
+    Properties:
+      NewInstancesProtectedFromScaleIn: true
+      VPCZoneIdentifier: !Split
+        - ','
+        - Fn::ImportValue: !Sub "${VPCStackParameter}-PrivateSubnetIds"
+      LaunchConfigurationName: !Ref 'ArmASGLaunchConfiguration'
+      MinSize: '0'
+      MaxSize: !Ref 'MaxSize'
+      DesiredCapacity: !Ref 'DesiredCapacity'
+      Tags:
+        - Key: Name
+          Value: !Sub 'ARM64-${ECSCluster}'
+          PropagateAtLaunch: true
+    CreationPolicy:
+      ResourceSignal:
+        Timeout: PT15M
+    UpdatePolicy:
+      AutoScalingReplacingUpdate:
+        WillReplace: true
+
+#Capacity Provider configuration to create CapacityProvider for ARM64 ASG. Capacity Provider needs ARM ASG Arn, 
+# so CloudFormation customer resource ARMClusterResource will make describe API call to ARM ASG to get the desired value. 
+  ArmECSCapacityProvider:
+    Type: AWS::ECS::CapacityProvider
+    Properties:
+        AutoScalingGroupProvider:
+            AutoScalingGroupArn: !Select [0, !GetAtt ARMCustomResource.Data ]
+            ManagedScaling:
+                MaximumScalingStepSize: 10
+                MinimumScalingStepSize: 1
+                Status: ENABLED
+                TargetCapacity: 100
+            ManagedTerminationProtection: ENABLED
+
+# ECS Task Definition for ARM64 Instance type. PlacementConstraints properties are setting the desired cpu-architecture to arm64.
+  Arm64taskdefinition:
+    Type: AWS::ECS::TaskDefinition
+    Properties:
+      Family: !Join ['', [!Ref 'AWS::StackName', -arm64]]
+      PlacementConstraints: 
```

Comment:
Can you add a description of this particular section in the task definition. It's important to explain here during the code review what we are doing by setting these constraints and why we are doing it. Same with the GPU section below.

Reply:
Fixed.